pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTValue.java

```diff
@@ -31,7 +31,7 @@
  * @author Andrew Bayer
  * @author Kohsuke Kawaguchi
  */
-public abstract class ModelASTValue extends ModelASTElement implements ModelASTMethodArg {
+public abstract class ModelASTValue extends ModelASTElement implements ModelASTMethodArg, ModelASTEnvironmentValue {
```

Comment:
`ModelASTEnvironmentValue` (and friends like `ModelASTWhenContent` and `ModelASTMethodArg`) are just marker interfaces - saying "any type implementing this interface can be used as a value where this interface is specified as the type". So `ModelASTValue`, for example, implements both `ModelASTEnvironmentValue` and `ModelASTMethodArg` because it can be used in both places (as well as any place that explicitly calls out `ModelASTValue`), but `ModelASTInternalFunctionCall` only implements `ModelASTEnvironmentValue` - so it can be used in contexts looking for `ModelASTEnvironmentValue`s but not those looking for `ModelASTMethodArg`, or any context that's explicitly looking for a `ModelASTValue`.

Thinking out loud, albeit a bit pedantically, as I try to make sure I understand what I'm saying... =) Imagine `Human`, `Lizard`, and `Cat` classes, with interfaces `HasTail` and `Mammal`. `Lizard` and `Cat` (ignore Manx cats for the moment!) would have the `HasTail` interface, while `Human` and `Cat` would have the `Mammal` interface (yes, not a *great* example since `Human` and `Cat` could just inherit from a `Mammal` class, but roll with me here).